test/lib/sdkdriver/src/dmp.ts

```diff
@@ -0,0 +1,132 @@
+ *   response.
+ */
+class DMPRequest {
+  private readonly _id: number;
+  private readonly _transport: Transport;
+  public readonly data: any;
+
+  /**
+   * Constructor.
+   * @param {number} id - Request ID
+   * @param {*} data - Request data
+   * @param {Transport} transport - The {@link Transport} used for
+   *   sending a response
+   */
+  constructor(id: number, data: any, transport: Transport) {
+    this._id = id;
+    this._transport = transport;
+    this.data = data;
+  }
+
+  /**
+   * Send a response over the underlying {@link Transport}.
+   * @param {*} data - Response data
+   * @returns {void}
+   */
+  sendResponse(data: any): void {
+    const { _id: id, _transport: transport } = this;
+    transport.send({ data, id, type: 'response' });
+  }
+}
+
+/**
+ * Driver Messaging Protocol (DMP).
+ * @class
+ * @classdesc A {@link DMP} exchanges DMP messages with its counterpart;
+ *   It can be instantiated in the mocha.js process that drives the test
+ *   process as well as in the test process
+ * @fires DMP#event
+ * @fires DMP#request
+ */
+export default class DMP extends EventEmitter {
+  private readonly _transport: Transport;
+  private _requestId: number;
+
+  /**
+   * Constructor.
+   * @param {Transport} transport - The {@link Transport} used for
+   *   exchanging DMP messages.
+   */
+  constructor(transport: Transport) {
+    super();
+    this._transport = transport;
+    this._requestId = 0;
+
+    this._transport.on('message', message => {
+      const { data, id, type } = message;
+      const transport: Transport = this._transport;
+
+      switch (type) {
+        case 'request':
+          this.emit('request', new DMPRequest(id, data, transport));
+          break;
+        case 'event':
+          this.emit('event', data);
+          break;
+      }
+    });
+  }
+
+  /**
+   * Close the underlying {@link Transport}.
+   * @returns {void}
+   */
+  close(): void {
+    this._transport.close();
+  }
+
+  /**
+   * Send an event over the underlying {@link Transport}.
+   * @param {*} data - Event data
+   * @returns {void}
+   */
+  sendEvent(data: any): void {
+    this._transport.send({
+      data,
+      type: 'event'
+    });
+  }
+
+  /**
+   * Send a request over the underlying {@link Transport}.
+   * @param {*} data - Request data
+   * @returns {Promise<*>} - Resolves with the response data
+   */
+  sendRequest(data: any): Promise<any> {
+    return new Promise((resolve, reject) => {
+      const requestId: number = this._requestId++;
+
+      const onresponse: (response: any) => void = response => {
+        const { id, type } = response;
+        if (type === 'response' && id === requestId) {
+          this._transport.removeListener('message', onresponse);
+          resolve(response.data);
+        }
+      };
+
+      this._transport.send({ data, id: requestId, type: 'request' });
+      this._transport.on('message', onresponse);
```

Comment:
There's one potential issue with this: the moment I have 10 outstanding requests or so, by default, Node's EventEmitter is going to warn about a potential leak. We can override this with [`setMaxListeners`](https://nodejs.org/api/events.html#events_emitter_setmaxlisteners_n)